### PR TITLE
CE create & update requests validation added

### PR DESCRIPTION
**Pull request by jdziurzynski**
KNTK-401

Implemented validation for create and update CloudExport API requests.

### CODE REVIEW

**Review by mateuszmidor — APPROVED** [review]

File: kentikapi/internal/validation/validate.go [hunk 1]
```diff
@@ -83,3 +85,81 @@ func validate(method string, direction string, path string, v reflect.Value) []s
 	}
 	return missing
 }
+
+// ValidateCECreateRequest checks if CloudExport create request contains all required fields.
+//nolint:gocyclo
```
Comment (danielfurman) [inline kentikapi/internal/validation/validate.go]: Please extract validateCEProperties function to increase readability. Unless you think that the cyclomatic complexity threshold is too low.

File: kentikapi/internal/validation/validate.go [hunk 2]
```diff
@@ -83,3 +85,81 @@ func validate(method string, direction string, path string, v reflect.Value) []s
...
+func ValidateCECreateRequest(ce *models.CloudExport) error {
+	if ce == nil {
+		return nil
```
Comment (danielfurman) [inline kentikapi/internal/validation/validate.go]:
```suggestion
		return fmt.Errorf("cloud export object is nil")
```
This should be an error, because there is no point in sending empty requests. Please also modify "nil request" subtest, of course.

File: kentikapi/internal/validation/validate.go [hunk 3]
```diff
@@ -83,3 +85,81 @@ func validate(method string, direction string, path string, v reflect.Value) []s
 	}
 	return missing
 }
+
```
Comment (danielfurman) [inline kentikapi/internal/validation/validate.go]: Please do not mix-up resource-agnostic validation helpers with resource-specific validation code. I suggest to move cloud export validation to cloud_export.go file.

File: kentikapi/internal/validation/validate.go [hunk 4]
```diff
@@ -83,3 +85,81 @@ func validate(method string, direction string, path string, v reflect.Value) []s
...
+			}
+		}
+	}
+	return nil
+}
+
+// ValidateCEUpdateRequest checks if CloudExport update request contains all required fields.
+func ValidateCEUpdateRequest(ce *models.CloudExport) error {
+	if ce == nil {
+		return nil
```
Comment (danielfurman) [inline kentikapi/internal/validation/validate.go]:
```suggestion
		return fmt.Errorf("cloud export object is nil")
```

File: kentikapi/internal/validation/validate.go [hunk 5]
```diff
@@ -83,3 +85,81 @@ func validate(method string, direction string, path string, v reflect.Value) []s
...
+			if ce.GetIBMProperties().Bucket == "" {
+				return ceFieldError("Properties.Bucket")
+			}
+		}
+	}
+	return nil
+}
+
+// ValidateCEUpdateRequest checks if CloudExport update request contains all required fields.
+func ValidateCEUpdateRequest(ce *models.CloudExport) error {
```
Comment (danielfurman) [inline kentikapi/internal/validation/validate.go]: Is update request with missing fields that are required for create (e.g. Name, PlanID, Properties.Bucket) accepted by the Kentik API?

**Review by danielfurman — CHANGES_REQUESTED** [review]

File: kentikapi/internal/validation/validate.go (hunk 5 above)
Comment (jdziurzynski) [reply kentikapi/internal/validation/validate.go]: You're right. I added the missing validation.

**Review by mmac-m3a — APPROVED** [review]
LGTM

**Comment by mmac-m3a** [comment]
You need to fix the failing test, before the code can be merged.

File: kentikapi/cloud_exports_integration_test.go [hunk 6]
```diff
@@ -523,13 +523,12 @@ func TestClient_CreateCloudExport(t *testing.T) {
 				assert.NoError(t, err)
 			}
 
-			if assert.Equal(t, 1, len(server.requests.createCERequests), "invalid number of requests") {
+			if tt.request != nil && assert.Equal(t, 1, len(server.requests.createCERequests), "invalid number of requests") {
```
Comment (danielfurman) [inline kentikapi/cloud_exports_integration_test.go]:
```suggestion
			if tt.expectedRequest != nil && assert.Equal(t, 1, len(server.requests.createCERequests), "invalid number of requests") {
```
Little more readable. If there is no expected request, do not verify it.
Also, the name of the test "nil request, status InvalidArgument received" should be changed to e.g. "nil request", because no request is performed in that case. Also, `tt.response` should be removed there, because it is not needed and we want to keep code lean.

File: kentikapi/internal/api_payloads/cloud_export.go [hunk 7]
```diff
@@ -251,3 +258,75 @@ func boolProtoPtrToBoolPtr(v *wrapperspb.BoolValue) *bool {
...
+	return validateCEProvider(ce)
+}
+
+//nolint:gocyclo
+func validateCEProvider(ce *models.CloudExport) error {
+	switch ce.CloudProvider {
+	case "":
+		return ceFieldError("CloudProvider")
+	case awsProvider:
+		if ce.GetAWSProperties().Bucket == "" {
```
Comment (danielfurman) [inline kentikapi/internal/api_payloads/cloud_export.go]:
```suggestion
                if ce.GetAWSProperties() == nil {return ceFieldError("Properties") }
		if ce.GetAWSProperties().Bucket == "" {
```
GetAWSProperties() might return nil pointer and nil pointer dereference causes panic. We need to do nil check to avoid that.

File: kentikapi/internal/api_payloads/cloud_export.go [hunk 8]
```diff
@@ -251,3 +258,75 @@ func boolProtoPtrToBoolPtr(v *wrapperspb.BoolValue) *bool {
 	}
 	return pointer.ToBool(v.GetValue())
 }
+
+// ValidateCECreateRequest checks if CloudExport create request contains all required fields.
+func ValidateCECreateRequest(ce *models.CloudExport) error {
```
Comment (danielfurman) [inline kentikapi/internal/api_payloads/cloud_export.go:263]: Please add a test for create request validation and update request validation. It might be a subtest in integration tests, that we already have. I suggest to test a missing AWS.Bucket field both for update and create. I know that testing every missing field is an overkill, but we should verify that the code works.

**Review by danielfurman — CHANGES_REQUESTED** [review]

File: kentikapi/internal/validation/validate.go (hunk 1 above)
Comment (danielfurman) [reply kentikapi/internal/validation/validate.go]: I still see `nolint` statement in the code, so the comment is still valid. Please extract more functions or let's discuss increasing gocyclo thershold.

File: kentikapi/internal/validation/validate.go (hunk 2 above)
Comment (danielfurman) [reply kentikapi/internal/validation/validate.go]: Done. Added separate comment for "nil request" subtest

File: kentikapi/internal/validation/validate.go (hunk 3 above)
Comment (danielfurman) [reply kentikapi/internal/validation/validate.go]: Done. Moving validation code to api_payloads package is acceptable for me, altough my comment didn't suggest that.

File: kentikapi/internal/validation/validate.go (hunk 4 above)
Comment (danielfurman) [reply kentikapi/internal/validation/validate.go]: Done

File: kentikapi/internal/validation/validate.go (hunk 5 above)
Comment (danielfurman) [reply kentikapi/internal/validation/validate.go]: Done

File: kentikapi/internal/api_payloads/cloud_export.go [hunk 9]
```diff
@@ -290,39 +290,71 @@ func ValidateCEUpdateRequest(ce *models.CloudExport) error {
...
 
-//nolint:gocyclo
 func validateCEProvider(ce *models.CloudExport) error {
 	switch ce.CloudProvider {
 	case "":
 		return ceFieldError("CloudProvider")
 	case awsProvider:
-		if ce.GetAWSProperties().Bucket == "" {
-			return ceFieldError("Properties.Bucket")
+		if err := validateAWSProvider(ce); err != nil {
```
Comment (danielfurman) [inline kentikapi/internal/api_payloads/cloud_export.go]:
```suggestion
		return validateAWSProvider(ce)
```
Minor simplification

File: kentikapi/internal/api_payloads/cloud_export.go [hunk 10]
```diff
@@ -290,39 +290,71 @@ func ValidateCEUpdateRequest(ce *models.CloudExport) error {
...
+			return err
 		}
 	case ibmProvider:
-		if ce.GetIBMProperties().Bucket == "" {
-			return ceFieldError("Properties.Bucket")
+		if err := validateIBMProvider(ce); err != nil {
+			return err
 		}
+	default:
+		return fmt.Errorf("CloudExport provider '%s' is not supported", ce.CloudProvider)
```
Comment (danielfurman) [inline kentikapi/internal/api_payloads/cloud_export.go]:
```suggestion
		return fmt.Errorf("cloud provider '%s' is not supported", ce.CloudProvider)
```

File: kentikapi/internal/api_payloads/cloud_export.go (hunk 7 above)
Comment (danielfurman) [reply kentikapi/internal/api_payloads/cloud_export.go]: This comment also applies to other providers, e.g. Azure

File: kentikapi/internal/validation/validate.go (hunk 1 above)
Comment (danielfurman) [reply kentikapi/internal/validation/validate.go]: Done

File: kentikapi/cloud_exports_integration_test.go (hunk 6 above)
Comment (danielfurman) [reply kentikapi/cloud_exports_integration_test.go]: Done

File: kentikapi/internal/api_payloads/cloud_export.go (hunk 8 above)
Comment (danielfurman) [reply kentikapi/internal/api_payloads/cloud_export.go:263]: Done

**Review by danielfurman — CHANGES_REQUESTED** [review]
As soon as we remove possible panics from functions validating properties it's good to go

File: kentikapi/internal/api_payloads/cloud_export.go (hunk 7 above)
Comment (danielfurman) [reply kentikapi/internal/api_payloads/cloud_export.go]: Done

File: kentikapi/internal/api_payloads/cloud_export.go (hunk 9 above)
Comment (danielfurman) [reply kentikapi/internal/api_payloads/cloud_export.go]: Done

File: kentikapi/internal/api_payloads/cloud_export.go (hunk 10 above)
Comment (danielfurman) [reply kentikapi/internal/api_payloads/cloud_export.go]: Done

**Review by danielfurman — APPROVED** [review]
Nice
